subsys/bluetooth/host/scan.c

```diff
@@ -639,22 +657,50 @@ void bt_hci_le_per_adv_report(struct net_buf *buf)
 	info.addr = &per_adv_sync->addr;
 	info.sid = per_adv_sync->sid;
 
-	if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL) {
-		/* Handling of incomplete reports is currently not
-		 * handled in the host. The remaining advertising
-		 * reports may therefore contain partial data.
-		 */
-		BT_WARN("Incomplete per adv report");
-	}
-
-	SYS_SLIST_FOR_EACH_CONTAINER(&pa_sync_cbs, listener, node) {
-		if (listener->recv) {
-			net_buf_simple_save(&buf->b, &state);
-
-			buf->len = evt->length;
-			listener->recv(per_adv_sync, &info, &buf->b);
-
-			net_buf_simple_restore(&buf->b, &state);
+	if (!per_adv_sync->report_truncated) {
+		if (per_adv_sync->reassembly == NULL) {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE) {
+				/* No reassembly buffer and a complete report.
+				 * Forward it without an extra copy.
+				 */
+				bt_hci_le_per_adv_report_recv(per_adv_sync, buf, &info);
+			} else {
+				BT_WARN("No reassembly buffer. Dropping the report");
+				per_adv_sync->report_truncated = false;
+			}
+		} else {
+			if (evt->data_status == BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_COMPLETE &&
+			    per_adv_sync->reassembly->len == 0) {
+				/* We have not received any partial data before.
+				 * This buffer can be forwarded without an extra copy.
+				 */
+				bt_hci_le_per_adv_report_recv(per_adv_sync, buf, &info);
+				net_buf_reset(per_adv_sync->reassembly);
+			} else {
+				size_t length = MIN(evt->length,
+						    net_buf_tailroom(per_adv_sync->reassembly));
+				net_buf_add_mem(per_adv_sync->reassembly, buf->data, length);
+				if (length < evt->length) {
+					/* The buffer is too small for the entire report. Drop it */
+					BT_WARN("Buffer is too small to reassemble the report. "
+						"Use CONFIG_BT_PER_ADV_SYNC_BUF_SIZE to change "
+						"the buffer size.");
+
+					per_adv_sync->report_truncated = true;
+					net_buf_reset(per_adv_sync->reassembly);
+					return;
+				}
```

Comment:
If we check lengths and discard events, then `MIN` value from line 680 doesn't provide anything useful. 

We should check if the event fit in the buffer first: if it does, copy it; if it does, do not copy any data (i.e. line 682 should only be done if the full report fit in the buffer).

Reply:
you are right, that is a leftover from when truncated reports were forwarded. I'll change the order